### ISSUE TITLE
Hilight problematic short segments when mouse hovers node

### ISSUE THREAD

**Issue by aubergine10** [CitiesSkylinesMods/TMPE#1229]
### Describe your idea

Very short segments cause vehicle AIs to behave unexpectedly when connected to nodes customised with the following tools:

* Priority Signs - #862 , #686
* Timed Traffic Lights - #302 , #297 , #862
* Junction Restrictions (probably - eg. enter blocked junction)

Those tools all require user to first select a node to edit before they can customise it; we can warn user about short node prior to selecting the node:

1. When user hovers mouse over the node they want to edit, check for short segments.
2. If found:
    * Highlight short segment(s) with red outline
    * Change mouse-following tooltip to "Tiny segments cause problems" (or something like that)
    * If not too painful, change mouse cursor to a warning symbol ⚠️ 
3. When mouse leaves node, or node selected for editing, clear the warning

See also: #698 

### Screenshots?

n/a

### Notes or questions?

Broken Nodes Detector mod has code that scans for short segments; could use that code but only check segments attached to the hovered node to reduce overheads.

Code for checking the segments attached to a node, and showing/clearing the warning could be centralised making it easier to incorporate in to any tools that need it. If memory serves, all those tools change the node highlight when mouse is over the node, so it would just be a case of adding maybe 1 line of code to the tools' node hover/leave/select handling methods.

The class should probably cache the node it just checked (and the result) so that it doesn't need to do expensive checks every frame. The cache would be cleared when the mouse leaves a node, or the node is selected for editing.

User could still select the node and customise it if they want to take the risk, but at least now they'll know what's causing problems if vehicles are ignoring their customisations.

Additional future enhancements:

* If we learn of other problems, eg. sometimes weird stuff like 'overlapping nodes' (user didn't properly connect two segments together), the central class could be updated to check for that too.
* A "Try and fix short segments" mod option; it's confirmed that reducing speed of a short segment can sometimes mitigate problems, at least for TTL. See also: #757

**Comment by krzychu124** [comment]
Honestly I will see is as a new overlay/icon layer just for warnings. Could be enabled and filtered based on active tool or something. Similar to what we see when one-way has a dead end, trolleybus/tram segment end is not connected, etc.

**Comment by krzychu124** [comment]
I've reviewed the code and found one big bug 😱 or I would say huge design flaw that makes junction features to work only when segment is long enough... segment length doesn't really matter if you know what you are doing and how simulation works (yes I've spent few days figuring out how simulation works in tiny details while working on EVE revival and improvements recently).

Test case (impossible to handle correctly running any version of the mod - yes, they have green here but as described, it doesn't matter - try yourself) :

https://user-images.githubusercontent.com/19638970/146283959-e47d6544-46bc-4110-a2b5-d8bf4a110529.mp4

Here the same scenario but with one little change in the code:

https://user-images.githubusercontent.com/19638970/146284048-2d7530b8-f122-46e5-8e2f-44e0b7125262.mp4

Bonus points: 
Yield signs are way more accurate, vehicle will wait patiently for perfect spot before start moving, no longer crawling to forcefully join with the traffic.

Big downside of the change: 
Makes whole implementation a lot slower, no fallback behavior, no shortcuts, no testing previous state, everything is calculated each time vehicle approach new segment. 

**Important note:** 
---
_Game can process even multiple segments/nodes per each simulation step for each vehicle in chunk (1024/step)_, number of processed segments is correlated with vehicle speed, speed limit and segments length. 
In above scenario I'm pretty sure it processed all 10 short segments in one step, then... 7-9 remaining, depends on how far vehicle was moved between steps and so on 😉

**Comment by aubergine10** [comment]
> Big downside of the change:
> Makes whole implementation a lot slower, no fallback behavior, no shortcuts, no testing previous state, everything is calculated each time vehicle approach new segment.

How much impact is there in relation to "a lot slower"? Is there some way we could embed "junction coming up, enable extra checks" in to the path units? That way the vehicle AI isn't doing all the extra work while driving down long stretches of road, it's only doing heavy lifting a few segments from a junction?

**Comment by krzychu124** [comment]
Each simulation step vehicle calculates 4 next positions (circles). Between steps it travels between current and next frame (rectangles) not further than to the first target, in the next step target0 is usually "converted" to "next" frame if speed is constant.

**130 km/h** (cur. segment is the segment of current pathUnit)

![image](https://user-images.githubusercontent.com/19638970/146297222-a6d0cce1-acfc-4519-84f5-5125c8d7209b.png)
![image](https://user-images.githubusercontent.com/19638970/146294842-358525e9-f118-4c59-9e61-b464257e859c.png)

**40 km/h**

![image](https://user-images.githubusercontent.com/19638970/146296160-0bb40161-6ad1-4ed2-a978-f30389bec4f4.png)
![image](https://user-images.githubusercontent.com/19638970/146296123-99224b0a-a429-48e2-9369-d866050dad84.png)

That change shouldn't have big impact for highways since segments are pretty long and also roads where junction node connects only two segments. I need to do some benchmarking, but keeping in mind user "cried" that simulation accuracy at very high was forced some time ago, this will be like multiple times of it if you place a lot of short segments and set medium speed (50-70km/h) with either traffic lights or priority signs.

**Comment by aubergine10** [comment]
Could it scale based on speed? So slower speed = less look-ahead?

**Comment by krzychu124** [comment]
> Could it scale based on vehicle speed? So slower vehicle speed = less look-ahead?

Distance between target positions is dynamic, based on vehicle speed, ability to brake and some other minor details.

> EDIT: Also, are those diagrams rendered in game (the circles/squares, not the text/arrows heh)?

It's a part of my Pathfind Sandbox (will be). I found nice and clean way to select/follow instance you want to observe so I started attaching more stuff to it in form of overlays.

**Comment by aubergine10** [comment]
Regarding EVE, there might be some possibility to use `Vehicle.Flags`? Specifically, the [`m_setvehicleflags`](https://cslmodding.info/asset/network/#m_setvehicleflags) of a segment is transposed on to vehicles driving on it... As this is only for roads, you could repurpose existing flags that are used for other AIs, such as `TakingOff`, `Flying`, `Landing`...?  Also I see there's `Emergency1` and `Emergency2` flags - no idea what they are used for (not checked).

**Comment by krzychu124** [comment]
> Congestion flag

Used in vanilla to force despawn vehicle approaching congested segment to prevent making longer traffic jams XD
> Emergency2

Used when vehicle use emergency signal lights
> Emergency1

Used in special cases, e.g. when fire truck stopped near the burning building and is extinguishing fire, similar for the fire copter, maintenance trucks when they maintain road, water trucks when pumping water, disaster response units searching for "problems" etc.
> m_setvehicleflags

Is applied unconditionally to every vehicle that enter segment, problem is it is per netInfo, so for all segment instances on the map. Good to help with deciding if effects like dust should be rendered, but not that useful for EVE (need to think about that). I've been testing different approach with better results.

**Comment by aubergine10** [comment]
Created wiki article to collate infos about `Vehicle.Flags`: https://github.com/CitiesSkylinesMods/TMPE/wiki/Vehicle-Flags

**Comment by krzychu124** [comment]
I was experimenting with priority signs code and I think we could improve some behaviour, especially when vehicle is driving off main road (yield other vehicles incoming from the front). 
Currently if junction does not have priority signs set or target and incoming vehicle are on main road they completely ignore a chance of possible collision. 

_Default behaviour, they will run into each other, ignoring priorities_

![image](https://user-images.githubusercontent.com/19638970/152698218-4815bd78-03ba-4d76-8c1f-289126ada86a.png)

_One way to fix it, but may introduce problem/unwanted yield for traffic going straight_ Blue truck will patiently wait for enough space or a gap in incoming traffic because of congested exit of intersection

![image](https://user-images.githubusercontent.com/19638970/152698232-6d3784f6-4aa8-4acf-ac09-b8b3c4043d7c.png)

What I would like to see: 
- blue truck should yield always when turning left (crossing incoming traffic), even if there are no priority signs
It might be active only when **Simulation Accuracy** is _High_ or _Very High_

Probably there are other cases where vehicles should yield. I noticed in code that once vehicle enter intersection area only vehicle collision is calculated - no further priority checks, so at big intersections or roundabouts it increases a chance of causing fake congestion on main road (going around the roundabout) blocking the traffic (yielding traffic will start entering intersection because of traffic gap mentioned above)

Maybe I'm blind, do we have any thread/issue with regards to problems/suggestions about priorities/priority signs, similar to those for traffic lights, parking AI etc.?

**Comment by aubergine10** [comment]
I remember some discussion of this, give me a few mins and I'll see if I can find it.

**Comment by aubergine10** [comment]
Note some of these aren't directly on topic but I think still worthy of considering as part of the mix:

* https://github.com/CitiesSkylinesMods/TMPE/issues/1134
* https://github.com/CitiesSkylinesMods/TMPE/issues/910
* https://github.com/CitiesSkylinesMods/TMPE/issues/1000
* https://github.com/CitiesSkylinesMods/TMPE/issues/871
* https://github.com/CitiesSkylinesMods/TMPE/issues/565
* https://github.com/CitiesSkylinesMods/TMPE/issues/424
* https://github.com/CitiesSkylinesMods/TMPE/issues/46

**Comment by aubergine10** [comment]
It would be super-useful if we could set priority by lane, rather than segment. That way the far-side turn lane could be given a stop/yield and the straight-on lanes could be priority to allow smooth traffic flow.

There's some other stuff relating to T-junctions and Y-junctions (forks) that needs considering which I don't think is directly mentioned in the linked issues above, I'll try and jot down some thoughts on those in an hour or so.

**Comment by krzychu124** [comment]
> It would be super-useful if we could set priority by lane, rather than segment. That way the far-side turn lane could be given a stop/yield and the straight-on lanes could be priority to allow smooth traffic flow.

This is what I meant, but for all cases without even touching intersection with TM:PE settings. **Traffic Priority** feature should be ON at _Maintenance_ tab, Sim Accuracy at least _High_ since checking priorities at every intersection may cause small slowdown (maybe check only in 300-500m from camera or reuse 0,1,2 LOD levels of vehicle simulation)


https://user-images.githubusercontent.com/19638970/152700929-041caea3-6e15-493c-9af9-067901eff3e8.mp4

**Comment by aubergine10** [comment]
that's freaking awesome!

**Comment by krzychu124** [comment]
Priorities are not tested at node so they behave a bit rough,

https://user-images.githubusercontent.com/19638970/152701220-1b02875a-881d-4c88-88ff-3fc03474bcc1.mp4

**Comment by aubergine10** [comment]
That's still a good improvement on what currently happens IMO

In terms of simulation workload, what's the "LOD level of vehicle simulation"? I assume the game is reducing complexity based on how far away the camera is?

**Comment by krzychu124** [comment]
> In terms of simulation workload, what's the "LOD level of vehicle simulation"? I assume the game is reducing complexity based on how far away the camera is?

Yes there are 3 LOD levels, at level 2 all collision checks are skipped, at level 1 only vehicle-pedestrian iirc

**Comment by krzychu124** [comment]
More "unprotected" turns.

Scenario:
- left going up,
- up going down,
- down going left,

https://user-images.githubusercontent.com/19638970/152716744-6a4fdd9f-4fe1-45d3-a2d2-6b410ed4c923.mp4

https://user-images.githubusercontent.com/19638970/152716729-9986994a-a622-4c20-ab7b-c8ac5cdebf69.mp4

https://user-images.githubusercontent.com/19638970/152716712-0aa5f885-f101-4da7-b79c-27f383f8b977.mp4

**Comment by aubergine10** [comment]
Looks really good!

**Comment by aubergine10** [comment]
BTW, will this also work for tracked vehicles? I have some nightmares in my current city with trains driving through each other lol

**Comment by krzychu124** [comment]
A bit more adv. setup.
In addition to previous, right is going left and right is going up

https://user-images.githubusercontent.com/19638970/152717810-95f78de2-0385-4efd-b417-efd5cca25b72.mp4

https://user-images.githubusercontent.com/19638970/152717822-d93f811f-605f-4a29-811d-81f8b119c7dc.mp4

https://user-images.githubusercontent.com/19638970/152717833-0a33020d-a377-4c7d-89e4-8d4d815923a9.mp4

https://user-images.githubusercontent.com/19638970/152717834-33021b1d-48a9-4508-9f8d-c62240c59073.mp4

> BTW, will this also work for tracked vehicles? I have some nightmares in my current city with trains driving through each other lol

Good question. I'll add some trams to the loop :D

**Comment by aubergine10** [comment]
Functional priority signs seem to completely obsolete current functionality of `enter blocked junction`.

I still feel strongly that `enter blocked junction` should be discarded and replaced with `box junction` (ie. only enter junction if the exit lane is not blocked) as per #424 

If `enter blocked junction` was retained, it could instead allow vehicles to part-enter and wait inside the junction node until there is gap in traffic - similar to how the `AdvancedJunctionRule` mod worked - #46 :

https://www.youtube.com/watch?v=ICjrLoy6tFA

**Comment by krzychu124** [comment]
A bit less realistic prio signs but still handles perfectly

https://user-images.githubusercontent.com/19638970/152720737-5e266578-ac93-454f-bf29-56eadd94e808.mp4

Removed all priorities so they all have priority - need to fix deadlock issue #1134 😄 

https://user-images.githubusercontent.com/19638970/152720993-37fbe664-b6ca-46ea-9d3e-4983f65771aa.mp4

**Comment by aubergine10** [comment]
I guess also that tracked vehicles should always be given right of way over road vehicles? Additionally maybe buses/trolleys could gain some priority over other road vehicles?

**Comment by aubergine10** [comment]
For deadlock issues, some ideas:

- Smaller / slower roads should yield to faster / larger roads
- If that still doesn't solve it, road with most congestion/traffic should go first

As mentioned above we could also have pecking order based on `ExtVehicleType` - eg. emergency, then tracked, then road-PT, then services, then cargo, then passenger car?

**Comment by krzychu124** [comment]
Yeah, we should write down info about priorities somewhere, maybe in the wiki as upcoming feature then just implement missing bits. 
In above case giving auto higher-priority to the Tram would potentially help with deadlock but still needs to be fixed in case of other road vehicles. Honestly I was browsing history of changes and couldn't find anti-deadlock code mentioned in the old wiki

> Functional priority signs seem to completely obsolete current functionality of enter blocked junction.

Yes it does not make sense if you put priority signs. I would apply it automatically under the hood e.g. for transition nodes when e.g.: road with median is connected to basic road.

`ExtVehicleType` will be good start then I think we could try introducing temporary priority, so e.g. once deadlock occurs we select best segment to get higher priority temporarily (e.g. for 30/60 sim frames to unload traffic from one side

**Comment by krzychu124** [comment]
Yeah, I remember `Advanced junction rule`. I've got an brief idea how to implement it but I need to create basic version of priority check function for vehicles already at node or like you've mentioned in the `box junction` area. It will used by vehicles that are on road with priority (intersecting with vehicles also on priority segment) but degraded to yielding because of unprotected turn maneuver, so the number of possible combinations can be optimized to be as fast as possible keeping best accuracy.

Hehe, I'm watching how they move... it's so satisfying to look at 😄

**Comment by aubergine10** [comment]
For deadlock I also think what would happen IRL...

Usually bigger vehicles can muscle in (and are used to doing so if you've ever known a lorry driver lol - they still do it when driving cars), so things like service and transport. But also you could get things like reckless drivers pushing in, and people on way to work might be more likely to enter junction, etc. For cars in particular we could possibly have it affected by individual driving style feature? I still think individual driving style should be on by default and use the driver id as if it were a randomised bitmask defining traits of the driver, including how they behave at junctions (are they more passive or aggressive, etc).

**Comment by aubergine10** [comment]
Another thing with individual styles, some drivers might be more willing to let others go first - eg. if there's a big queue they might flash lights to let car at front enter junction (or exit junction if they were part entered).

**Comment by krzychu124** [comment]
Btw, now as I watch they wait for their turn I started thinking about fixing blinkers so even if vehicle does not move and e.g.: waiting to turn left it could have blink left ON (that's possible, just require a tiny patch).

> Usually bigger vehicles can muscle in (and are used to doing so if you've ever known a lorry driver lol - they still do it when driving cars), so things like service and transport

Yeah, I thought about that too. At worst we can select them by ID -> decision is taken every sim step so every 1024/4096(MV) vehicles - we can "push" first found vehicle from current chunk 🙃 
There are many ways to do so.

**Comment by aubergine10** [comment]
> Btw, now as I watch they wait for their turn I started thinking about fixing blinkers so even if vehicle does not move and e.g.: waiting to turn left it could have blink left ON (that's possible, just require a tiny patch).

Additionally, it would be great if they didn't use blinkers when simply driving round a curved segment (unless it's one-way street where it might be roundabout heh).

I have some really bendy roads snaking through mountains and it always irrates me how vehicles indicate round sharp bends lol.

**Comment by krzychu124** [comment]
Fully manual prototype

https://user-images.githubusercontent.com/19638970/152874006-bff2e88c-387b-4f56-b6ea-1fc55f87c816.mp4

**Comment by aubergine10** [comment]
Ooh, can we make it so if a vehicle breaks fast it puts both indicators on to warn other traffic?

**Comment by krzychu124** [comment]
> Ooh, can we make it so if a vehicle breaks fast it puts both indicators on to warn other traffic?

I have no idea how to detect it.

> Could also do both indicators when vehicles are pulled in to a building to load/unload/etc.

Will be the easiest thing actually 😄 

I was thinking about bus tops - emergency when boarding then turn to join traffic. Since most calls are invoked 4 times/per vehicle/per step it's not obvious where to put 'start' and 'reset' triggers 🤔

**Comment by aubergine10** [comment]
> Since most calls are invoked 4 times/per vehicle/per step

omg, wat?!

**Comment by krzychu124** [comment]
Movement is interpolated between frames (last and next - rectangles). Next calculation occurs when vehicle is about to hit next frame. It needs to calculate 4 target positions (circles), they are used as next position (target0, closest), other to determine if vehicle will accelerate or slow down/ stop. Most of TM:PE methods, like eg. `MayChangeSegment` - when calculated target position will land as last position of current segment, at node or next segment. `CalculateSegmentPosition` is called for each calculated target position (circle) that should land on segment, otherwise simplified version is called to apply transition on the node. Etc. It's not that card to understand how it works and what to do in case of tweaks/changes

5% speed:

https://user-images.githubusercontent.com/19638970/152890046-eb9a5be1-84a9-468d-987d-300d4edb465f.mp4


https://user-images.githubusercontent.com/19638970/152890408-19f59e93-c57b-459c-b5f9-6a03346d8666.mp4


Normal speed XD

https://user-images.githubusercontent.com/19638970/152890502-75e37327-80b3-40a9-8513-f51fe83bbdfd.mp4

**Comment by krzychu124** [comment]
More or less _blinking_😄 

https://user-images.githubusercontent.com/19638970/152911472-848ed960-3f4e-48c9-958c-a47c39a0c2bb.mp4

https://user-images.githubusercontent.com/19638970/152911550-d52c153e-32fe-46f2-ba4f-972a8f4e6f14.mp4

**Comment by aubergine10** [comment]
nice!

**Comment by krzychu124** [comment]
> How is the blinking done? Is it just a flag that gets set and the game does the rest, or do you have to write code to keep turning it on/off?

Currently I'm checking direction between current and next segment. If `left` or `right` then turn on turn signal otherwise reset. I need to add a little code to check if approaching node is a junction to eliminate turn signals on Bend nodes.
I wanted to add turn signals when changing lanes but I couldn't find performant solution for calculating if next lane changes and which direction compared to current movement direction. Also I couldn't find way to detect if a bus stop is on the left or right side...

**Comment by brunoais** [comment]
Can nodes store arbitrary metadata?
If so, maybe you can store in nodes when there's an intersection approaching in nodes nearby intersections (by distance, instead of by node count). I don't know how hard would it be to maintain such information for when the road changes. Maybe you can use the agents themselves to report if no intersection was expected and an intersection is encountered so that information can be added to the close-by nodes.

**Comment by aubergine10** [comment]
> I wanted to add turn signals when changing lanes but I couldn't find performant solution for calculating if next lane changes and which direction compared to current movement direction

The way lane changes are limited to nodes, and the speed at which vehicles just jolt in to next lane would make indicators look weird in that context anyway IMO.

It would be interesting to see the vehicle frames overlay (or whatever it's called) at 5% speed as a vehicle encounters a lane change.

> Also I couldn't find way to detect if a bus stop is on the left or right side..

Very basic solution would be based on which side of road traffic drives on for the current city - `Shortcuts.cs`:

```csharp
        internal static bool LHT => Singleton<SimulationManager>.instance.m_metaData.m_invertTraffic == SimulationMetaData.MetaBool.True;
        internal static bool RHT => !LHT;
```

But there are situations where bus stop is on a median, or even other side of road (in case of 1-way busways) - to discern that you'd need list of sorted lanes (there's already code for that) but ideally cached because it would be ludicrous to keep recalculating when segments very rarely change (geometry observer can help invalidate cache where necessary - this is also something desperately needed for overlays and which I'll probably make a start on after mod options chainsawing).

With cached list of sorted lanes for segment, it's simple case of checking if lane `0` (or lane `Count-1`?) in the lane that the bus enters the bus stop from.

Also would need to skip the whole thing if bus is stopping in a building (eg. bus terminal).

Actually, it doesn't matter if the code to determine which side the bus stop is on is bloaty and slow - it only needs to run once when the stop is created/moved. We need `ExtTransportStop` for buses, trams, trolleys :) lol

**Comment by aubergine10** [comment]
Oh, one other scenario for which side bus is on that breaks my earlier assertion that we can just check a single lane position...

Roads with central bus lanes separated from normal road lanes by median, eg: various BRT roads [like this one](https://steamcommunity.com/sharedfiles/filedetails/?id=1275797422) in workshop:

![image](https://user-images.githubusercontent.com/1386719/153043634-13ccc36a-ab10-44dc-b785-5420dd58b18b.png)

There are also [inverted versions of that](https://steamcommunity.com/sharedfiles/filedetails/?id=2069050333):

![image](https://user-images.githubusercontent.com/1386719/153043789-2e8afe85-db55-4405-920c-858f10a337ec.png)

**Comment by krzychu124** [comment]
> Very basic solution would be based on which side of road traffic drives on for the current city

I just need a way to say that lane is directly connected or connection requires changing to the left or right.

> The way lane changes are limited to nodes, and the speed at which vehicles just jolt in to next lane would make indicators look weird in that context anyway IMO.

Not at all. Just was clip with overlays - notice that gray lane marks current lane and I can add code for testing lane change at each circle (in the first clip after few sec. first yellow circle is calculating position on the next segment) so depends on speed I can turn on indicator even one or two segments before actual lane change occurs and even better, when congested you will see them blinking way before they actually change lane. It will be nice addition. I can also inject same thing into DLS code, so when TM:PE decide there is a better lane I can turn on signal in that case too - again segment or even multiple before lane change.
The main problem that needs to be solved and will basically fix all issues -> I need a function that for two laneIDs or lane indexes would tell me if lane transition involves lane change or not and if so, tell relative direction of it.

Notice how far vehicle needs to move after turning left till signal stops blinking. IMO, pretty nice.

https://user-images.githubusercontent.com/19638970/153064512-09233a4d-15ba-40be-a17b-1ce4ced40327.mp4

At 99% because of DLS but still, lane change is not instant. You can detect is pretty long ahead

https://user-images.githubusercontent.com/19638970/153065607-c79e22ae-7dbb-4171-9e1f-e6a1a77338d5.mp4
